DM-29795: Add meas_extensions_trailedSources to lsst_distrib

langfzac opened the pull request:
Implements a trailed source measurement extension ([RFC-768](https://jira.lsstcorp.org/browse/RFC-768)). This includes two measurement plugins: 
1) `SingleFrameNaiveTrailPlugin`: Measures the length, angle, and end points of extended sources using the second moments. 
2) `SingleFrameVeresTrailPlugin`: Measures the length, angle, end points, flux, and centroid of an extended source by fitting the trail model from [Vereš et al. 2012](https://ui.adsabs.harvard.edu/abs/2012PASP..124.1197V/abstract) (an axisymmetric Gaussian convolved with a line).

timj commented:
Please remember to add the lint.yaml github action (you should be able to find it in any of the standard packages).

kfindeisen commented on examples/SConscript:
```diff
@@ -0,0 +1,3 @@
+# -*- python -*-
+from lsst.sconsUtils import scripts
+scripts.BasicSConscript.examples()
```
I'd recommend against having an `examples/` directory, as these have proven rarely-used and difficult to maintain. Put code examples in the documentation instead.

kfindeisen commented on include/lsst/meas/extensions/trailedSources.h:
```diff
@@ -0,0 +1,31 @@
+
+/** \file
+  * \brief Include files for lsst::meas::extensions::trailedSources.
+  */
```
We [no longer use file documentation comments](https://developer.lsst.io/cpp/api-docs.html#boilerplate-for-header-h-and-source-cc-files), nor `\`-style Doxygen commands. It looks like the package template is badly out of date; I've opened [DM-29922](https://jira.lsstcorp.org/browse/DM-29922) to address that.

kfindeisen commented on ups/meas_extensions_trailedSources.cfg:
```diff
@@ -0,0 +1,25 @@
+config = lsst.sconsUtils.Configuration(
+    __file__,
+    headers=["lsst/meas/extensions/trailedSources.h"],
+    hasDoxygenInclude=False,
```
Need `hasSwigFiles=False` (SWIG was an old Python wrapping system system we no longer use).

kfindeisen commented on ups/meas_extensions_trailedSources.cfg:
```diff
@@ -0,0 +1,25 @@
+dependencies = {
+    "required": ["utils", "ndarray"],
+    "buildRequired": ["boost_test", "pybind11", "ndarray", "pex_exceptions",
+                      "pex_config", "afw", "geom", "meas_base"],
```
This needs to be synchronized with your actual dependencies in `meas_extensions_trailedSources.table`. At the very least, you don't need `boost_test`, as you never use it. Not sure about `pex_exceptions` (which you don't seem to use explicitly, but is also missing from the table file).

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+ */
+
+#ifndef LSST_MEAS_EXTENSIONS_TRAILEDSOURCES_VERES_H
+#define LSST_MEAS_EXTENSIONS_TRAILEDSOURCES_VERES_H
```
The [usual convention](https://developer.lsst.io/cpp/style.html#header-files-must-include-a-construction-that-prevents-multiple-inclusion) is:
```suggestion
#ifndef LSST_MEAS_EXTENSIONS_TRAILEDSOURCES_VERES_MODEL_H
#define LSST_MEAS_EXTENSIONS_TRAILEDSOURCES_VERES_MODEL_H
```

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+#include "lsst/afw/image.h"
+#include "lsst/afw/geom.h"
+#include "lsst/geom.h"
+#include "ndarray.h"
```
Sorting [by type and dependency](https://developer.lsst.io/cpp/style.html#include-statements-should-be-sorted-and-grouped) is preferred:
```suggestion
#include "ndarray.h"

#include "lsst/pex/config.h"
#include "lsst/geom.h"
#include "lsst/afw/geom.h"
#include "lsst/afw/image.h"
```

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+ * for a fast-moving, trailed-source (Veres et al. 2012).
+ *
+ * VeresModel is designed to compute the chi-squared of the model given an
+ * `lsst::afw::image::Exposure`, and to be passed to scipy.optimize for minimization.
```
The backticks will turn off Doxygen's link generation; leave them out in C++.

kfindeisen replied:
The reference to `scipy.optimize` seems oddly specific. If there something about this class that's specific to it, as opposed to other math and model-fitting libraries (e.g., follows a specific interface)? If so, please explain.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+class VeresModel {
+public:
+    typedef afw::image::Image<float> Image;
+    typedef afw::image::Exposure<float> Exposure;
```
Assuming these are intended as shorthands and not as abstractions, consider calling them `ImageF` and `ExposureF`, per [existing Pipelines convention](https://developer.lsst.io/cpp/style.html#a-a-name-representing-a-typedef-should-have-a-t-suffix-if-and-only-if-necessary-to-disambiguate-the-typedef-from-a-template-bare-name).

If these are shorthands, it might be worth making them private to avoid cluttering up the class documentation. However, I'm not sure what that would do the the docs for the method signatures.

kfindeisen replied:
The use of `typedef` is discouraged as of C++11; prefer `using` (I'll admit the style guide is not up to date on this):
```suggestion
    using Image = afw::image::Image<float>;
    using Exposure = afw::image::Exposure<float>;
```

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+ * VeresModel is designed to compute the chi-squared of the model given an
+ * `lsst::afw::image::Exposure`, and to be passed to scipy.optimize for minimization.
+ */
+class VeresModel {
```
Consider making the class `final` if it's not intended to have subclasses.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+     * @param data Exposure passed from the measurement task.
+     */
+    explicit VeresModel(Exposure const& data);
+
```
DM convention requires [explicitly declaring copy-constructors](https://developer.lsst.io/cpp/style.html#b-classes-should-explicitly-declare-or-delete-compiler-generated-members). Consider whether or not the default implementations give the desired behavior for `Image::Array`.

langfzac replied:
Ok. I've added these and everything compiles, but I'm not entirely sure how to think about/check the behavior of the default constructors...

kfindeisen replied:
The main issue was that a `ImageF::Array` (and LSST image-ish types in general) is designed for shallow copies, so the compiler-generated copy-constructor would create two `VeresModel` objects with shared state. Since I see the only remaining `Array` members are `_data` and `_variance`, which are supposed to be fixed at construction time, this is probably ok now.

langfzac replied:
Ah ok, so the (now removed) `_model` would've been updated inside of both the original and the 'copied' class? And to be sure, with that resolved, the default constructors (in the link above) will suffice?

kfindeisen replied:
They look reasonable to me; I think that each call to `operator()` or `computeGradient()` is independent enough that changes to other objects (including the original `Exposure`) won't introduce any problems. However, it's your class, and your decision on how it _should_ behave.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return The chi-squared of the model.
+     */
+    double operator()(std::vector<double> const& params) const;
```
Can this method be `noexcept`, or does it throw exceptions on invalid inputs or data? If the latter, please add one or more [`@throws`](https://developer.lsst.io/cpp/api-docs.html#cpp-doxygen-throws).

langfzac replied:
Yes, I think I should add some checks. Are there any stack-defined exceptions that should be used?

kfindeisen replied:
Yes, you should use subclasses of [`lsst::pex::exceptions::Exception`](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/classlsst_1_1pex_1_1exceptions_1_1_exception.html). I assume [`lsst::pex::exceptions::InvalidParameterError`](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/classlsst_1_1pex_1_1exceptions_1_1_invalid_parameter_error.html) is the appropriate one for this case, though in the context of minimization it's not really a logic error...

If you use any `pex::exceptions` exception, construct it with the [`LSST_EXCEPT` macro](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/_exception_8h.html#ae48de0285726e06656de9ee35c1530af), which adds things like line numbers.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return The gradient of chi-squared of the model.
+     */
+    std::vector<double> gradient(std::vector<double> const& params);
```
Can this method be `const` and/or `noexcept`?

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+     * Compute chi-squared of the model given the data.
+     *
+     * @param params Model parameters. [centroid x, centroid y, trail flux,
+     *     trail length, trail angle].
```
Are there any more details that are assumed by the model (units, zero point, angle convention, etc.)? If they're long, these may need to go in a notes section instead of `@param`.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+    std::vector<double> gradient(std::vector<double> const& params);
+
+    /// Return the PSF sigma.
+    double getSigma() const { return _sigma; }
```
This one can definitely be `noexcept`.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+
+    /// Computes the gradient of the Veres et al. 2012 model for a given pixel located at (x,y).
+    void _computeGradient(double x, double y, double xc, double yc,
+                         double F, double L, double theta);
```
Adding more thorough documentation will make this code easier to maintain in the future. In particular, if you decide to keep `_gradModel` (see below), the fact that `_computeGradient` has side effects on other private members should be documented explicitly.

kfindeisen replied:
We don't have a policy against  1- or 2-character parameter names, but please at least make them lower case, to avoid confusion [with constants](https://developer.lsst.io/cpp/style.html#named-constants-including-enumeration-values-must-be-all-uppercase-using-underscore-to-separate-words).

kfindeisen replied:
I recommend declaring both of these methods `noexcept`, since as I understand it they're *supposed* to be pure math, with no input validation, and no future edits are likely to change that.

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+    Image::Array _model;
+    Image::Array _data;
+    Image::Array _variance;
+    std::vector<double> _gradModel;
```
Consider documenting these with at least regular comments, since it's not obvious what all these things are for  without reading the code. In particular, it's not clear which of these are only set by construction/assignment and which ones are altered by model fitting.

kfindeisen replied:
I suggest removing `_model` and `_gradModel` entirely, since the code uses them to buffer information that could just as well be stored in local variables (e.g., the implementation of `operator()` writes to `_model[x, y]`, then immediately uses that value, and never reads the same pixel again, nor does any other method read `_model[x, y]` without recomputing it first).

If you're worried about the initialization overhead of returning a vector from `_computeGradient`, note that `std::array` is a lightweight alternative, and it won't affect the Python interface at all.

Removing these two members would make both the function implementations and the class structure much easier to reason about, since you don't have to argue about whether `operator()` is truly `const` or whether, as temporary storage, `_model` and `_gradModel` should be declared `mutable`. (Yes, `ndarray` lets you modify pixels through a reference to `const` array, but that's a design flaw in the library.) It will help with compiler optimization as well if it can see that these values are short-lived.

kfindeisen commented on python/lsst/meas/extensions/trailedSources/SConscript:
```diff
@@ -0,0 +1,2 @@
+from lsst.sconsUtils import scripts
+scripts.BasicSConscript.python(['_trailedSources'])
```
I believe you shouldn't need to pass an argument to `python()` unless you're using submodules. At least, it's not recommended by [the pybind11 tutorial](https://developer.lsst.io/pybind11/how-to.html#building-the-wrapper).

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+    std::vector<double> _gradModel;
+};
+
+}}}} // lsst::meas::extensions::trailedSources
```
Consider running [`clang-format`](https://developer.lsst.io/cpp/style.html#layout-may-be-automated-with-clang-format-using-the-lsst-configuration), if you have it available; that will make the code easier to maintain by developers who use it.

kfindeisen commented on python/lsst/meas/extensions/trailedSources/_VeresModel.cc:
```diff
@@ -0,0 +1,58 @@
+        [](auto & mod, auto & cls) {
+            cls.def(py::init<afw::image::Exposure<float> const&>(), "data"_a);
+            cls.def("__call__",
+                py::overload_cast<std::vector<double> const&>(&VeresModel::operator(), py::const_));
```
You don't need `py::overload_cast` or `py::const_` here, since there's only one `operator()`.

kfindeisen commented on python/lsst/meas/extensions/trailedSources/_VeresModel.cc:
```diff
@@ -0,0 +1,58 @@
+
+    wrappers.addSignatureDependency("lsst.geom");
+    wrappers.addSignatureDependency("lsst.afw.image");
+    wrappers.addSignatureDependency("lsst.afw.detection");
```
Where does the dependency on `afw.detection` come from? I don't see any detection code, and `VeresModel.h` doesn't include `detection.h`.

kfindeisen commented on python/lsst/meas/extensions/trailedSources/_VeresModel.cc:
```diff
@@ -0,0 +1,58 @@
+            cls.def(py::init<afw::image::Exposure<float> const&>(), "data"_a);
+            cls.def("__call__",
+                py::overload_cast<std::vector<double> const&>(&VeresModel::operator(), py::const_));
+            cls.def("gradient", &VeresModel::gradient);
```
Please [declare the parameters](https://developer.lsst.io/pybind11/style.html#keyword-names-must-be-provided-for-functions-with-multiple-overloads-or-more-than-two-arguments) for `__call__` and `gradient`; this will make Python-side docs (including error messages) clearer.

kfindeisen commented on python/lsst/meas/extensions/trailedSources/_VeresModel.cc:
```diff
@@ -0,0 +1,58 @@
+#include "lsst/meas/extensions/trailedSources/VeresModel.h"
+#include "lsst/geom.h"
+#include "lsst/afw/image.h"
+#include "lsst/afw/detection.h"
```
[Dependency order](https://developer.lsst.io/cpp/style.html#include-statements-should-be-sorted-and-grouped) would place `VeresModel.h` at the bottom.

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+#include "lsst/meas/extensions/trailedSources/VeresModel.h"
+#include "lsst/geom.h"
+#include "lsst/afw/image.h"
+#include "lsst/afw/detection.h"
```
Not in dependency order; I also don't think you use `detection.h` anywhere.

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+namespace trailedSources {
+
+typedef afw::image::Image<float> Image;
+typedef afw::image::Exposure<float> Exposure;
```
Same comments as in the header. Note that you could also use `VeresModel::Exposure`, though that's probably more appropriate if it's an abstraction rather than a shorthand.

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+    double xc = params[0];    // Centroid x
+    double yc = params[1];    // Centroid y
+    double F = params[2];     // Flux
+    double L = params[3];     // Trail length
```
Do not use all-uppercase for [local variables](https://developer.lsst.io/cpp/style.html#variable-names-must-be-camelcase-or-snake-case), it may confuse other readers.

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+    double L = params[3];     // Trail length
+    double theta = params[4]; // Angle from +x-axis
+
+    // From computeKernelImage()
```
I don't understand this comment. What is `computeKernelImage()`?

langfzac replied:
This is a personal comment I forgot to rewrite. It's supposed to say that the following loop was adapted from the `lsst::afw::detection::Psf::computeKernalImage()`. Is there a proper was to cite something like this?

kfindeisen replied:
Not that I know of, no. Your revised wording sounds clear enough. 🙂

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+    // From computeKernelImage()
+    // Compute model image and chi-squared
+    double chiSq = 0.0;
+    double tmp = 0.0;
```
It's better style (for safety, performance, and readability) to declare `tmp` only when you need it, in this case inside the inner loop. ~I don't have a copy of Effective C++ handy to give an exact citation, sorry.~ (Edit: FWIW, Effective C++, 3rd Ed., Item 26)

langfzac replied:
Not exactly sure what you mean. Are you saying the loop should look something like below? (With proper variable naming, of course)
```c++
double chiSq = 0.0;
for (;;) {
    for (;;) {
        double tmp = something;
        chiSq += tmp*tmp
    }
}
```

kfindeisen replied:
Assuming those aren't actually null loops, then yes.

Edit: also, exactly the same argument applies to the new `model` local -- if you don't need to persist state between loop iterations, give the variable as narrow a scope as possible.

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+        Image::Array::Reference varRow = _variance[yIndex];
+        for (int xIndex = 0, xp = _bbox.getBeginX(); xIndex < _bbox.getWidth(); ++xIndex, ++xp) {
+            modelRow[xIndex] = _computeModel(xp,yp,xc,yc,F,L,theta);
+            tmp = dataRow[xIndex] - modelRow[xIndex];
```
Consider a more informative name than `tmp`. `delta` or `diff`, maybe?

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+        for (int xIndex = 0, xp = _bbox.getBeginX(); xIndex < _bbox.getWidth(); ++xIndex, ++xp) {
+            modelRow[xIndex] = _computeModel(xp,yp,xc,yc,F,L,theta);
+            _computeGradient(xp,yp,xc,yc,F,L,theta);
+            tmp = -2.0 * (dataRow[xIndex] - modelRow[xIndex]) / varRow[xIndex];
```
More informative name?

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+
+void VeresModel::_computeGradient(double x, double y, double xc, double yc,
+                                  double F, double L, double theta) {
+    double xp = (x-xc)*std::cos(theta) + (y-yc)*std::sin(theta);
```
It might be worth adding a local `using namespace std` or `using std::cos, std::sin, std::sqrt, std::exp` (list syntax new with C++17) to cut down on all the `std::` qualifiers.

kfindeisen commented on src/VeresModel.cc:
```diff
@@ -0,0 +1,157 @@
+    double sigmaSq8 = sigmaSq * 8.0;
+    double eypSq =  std::exp(-ypSq/(2.0*sigmaSq));
+    double LPlus = L+2.0*xp;
+    double LMinus= L-2.0*xp;
```
Please [use camelCase](https://developer.lsst.io/cpp/style.html#variable-names-must-be-camelcase-or-snake-case).

kfindeisen commented on include/lsst/meas/extensions/trailedSources/VeresModel.h:
```diff
@@ -0,0 +1,100 @@
+     * @param params Model parameters. [centroid x, centroid y, trail flux,
+     *     trail length, trail angle].
+     *
+     * @return The gradient of chi-squared of the model.
```
Document that this is [d/dparam...]; I assumed it was [d/dx, d/dy] before realizing it doesn't make sense.

kfindeisen commented on setup.cfg:
```diff
@@ -8,6 +8,7 @@ exclude =
   **/*/__init__.py,
   **/*/version.py,
   tests/.tests
+  tests/test_trailedSources.py
```
Test code normally should be checked. Is there a reason to exclude this file, specifically?

kfindeisen replied:
It's a bit hard to change this now, but in the future, try to commit tests at the same time as the corresponding code. This makes the commits [more self-contained](https://developer.lsst.io/work/flow.html#commits-should-represent-discrete-logical-changes-to-the-code), and lets you catch problems like the missing dependencies right away.

kfindeisen reviewed (approved):
C++ and infrastructure look good! Since this is a new package, I erred on the side of thoroughness in reporting style issues. The biggest functional issue is that I think `VeresModel` could be made _much_ simpler, but there's no reason for that to block merging.

morriscb commented on python/lsst/meas/extensions/trailedSources/NaivePlugin.py:
```diff
@@ -0,0 +1,186 @@
+@register("ext_trailedSources_Naive")
+class SingleFrameNaiveTrailPlugin(SingleFramePlugin):
+    """Naive trailed source characterization plugin
+    Parameters
```
For readability of the docs here and throughout the python packages, add a blank line between the different sections (e.g. Parameters, Returns) in the doc strings.

morriscb reviewed (commented):
Documentation builds with only minor formatting warnings. Follow the instructions here: https://developer.lsst.io/stack/building-single-package-docs.html to build the docs and fix the formatting causing the warnings.

kgabor commented on tests/test_trailedSources.py:
```diff
@@ -0,0 +1,207 @@
+from lsst.utils.tests import methodParameters
+
+# Trailed-source length, angle, and centroid.
+np.random.seed(432)
```
Please use the newer generator API of np.random (https://numpy.org/doc/stable/reference/random/generator.html) here and in TrailedTestDataset as well.

kgabor commented on python/lsst/meas/extensions/trailedSources/VeresPlugin.py:
```diff
@@ -0,0 +1,188 @@
+        default="L-BFGS-B"
+    )
+
+    def setDefaults(self):
```
Empty setDefaults can be omitted.

kgabor commented on python/lsst/meas/extensions/trailedSources/NaivePlugin.py:
```diff
@@ -0,0 +1,186 @@
+    """Config class for SingleFrameNaiveTrailPlugin.
+    """
+
+    def setDefaults(self):
```
Empty `setDefaults()` can be omitted for clarity. The whole class definition becomes `pass`.

kgabor commented on python/lsst/meas/extensions/trailedSources/NaivePlugin.py:
```diff
@@ -0,0 +1,186 @@
+        F = measRecord.get("base_SdssShape_instFlux")
+
+        # Fall back to aperture flux
+        if np.isnan(F):
```
Here and at a few other places: Is it NaN only that is rejected and +/-Inf acceptable? If not, I'd recommend using `np.isfinite` instead.

kgabor commented on python/lsst/meas/extensions/trailedSources/NaivePlugin.py:
```diff
@@ -0,0 +1,186 @@
+        xmy2 = xmy*xmy
+        xy2 = Ixy*Ixy
+        a = np.sqrt(0.5 * (xpy + np.sqrt((xmy)**2 + 4.0*xy2)))
+        L = 2*a
```
`L = 2.0*a`
`(xmy)**2` -> already introduced `xmy2`

kgabor commented on python/lsst/meas/extensions/trailedSources/NaivePlugin.py:
```diff
@@ -0,0 +1,186 @@
+        L = 2*a
+
+        theta = 0.5 * np.arctan2(2.0 * Ixy, xmy)
+        x0 = xc-a*np.cos(theta)
```
Optimisation:  `a*np.sin(theta)` and  `a*np.cos(theta)` are the same as in L155-158. `dxdt = ` and `dydt =` can be moved here to avoid repeating calculations.

Style: Please use spaces around +,- operators.

kgabor commented on python/lsst/meas/extensions/trailedSources/NaivePlugin.py:
```diff
@@ -0,0 +1,186 @@
+moments. The angle, :math:`\Theta`, from the x-axis is computed similarly (via
+second moments). The end points of the trail are then given by
+:math:`(xc \pm a*cos(\Theta),yc \pm a*sin(\Theta))`.
+"""
```
Here and also in `VeresPlugin.py`: As this module implements one task only, I'd recommend moving this docstring into the class level docstring of `SingleFrameNaiveTrailPlugin`. The details can go into the `Extended summary` and `Notes` sections. 

https://developer.lsst.io/python/numpydoc.html?highlight=class%20docstrings#placement-of-module-docstrings

kgabor commented on python/lsst/meas/extensions/trailedSources/VeresPlugin.py:
```diff
@@ -0,0 +1,188 @@
+"""
+
+import numpy as np
+from scipy.optimize import minimize
```
It may be confusing to import `minimize` as a global function in this module. As it appears only once, please keep a longer form (`scipy.optimize.minimize()` or `sciOpt.minimize()` etc.)

kgabor commented on tests/test_trailedSources.py:
```diff
@@ -0,0 +1,207 @@
+# Taken from NaiveCentroidTestCase
+class TrailedSourcesTestCase(AlgorithmTestCase, lsst.utils.tests.TestCase):
+
+    @methodParameters(L=Ls, theta=thetas, xc=xcs, yc=ycs)
```
I don't think that this was the intended behavior: 
 - The decorated `setUp` is called right before `testNaivePlugin` (and `testVeresPlugin`).
 - `methodParameters` loop through the inputs, calling the original `setUp` for each input.
 - `testNaivePlugin` (and `testVeresPlugin`) actually _runs only once_, with the result of the very last `setUp` call.

langfzac replied:
This has been changed to use a `classParameters` decorator on the class itself. Thanks!

kgabor reviewed (changes requested):
I did not check whether all the (python) calculation formulas are algorithmically correct. I'm afraid, I have one major comment for the unit test.

langfzac commented:
Ok, I believe I've addressed all the comments. 

@kgabor The issue with the tests has been resolved.

@morriscb The documentation now builds with no warnings.
